core/src/main/scala/kafka/network/PrioritizationAwareBlockingQueue.scala

```diff
@@ -0,0 +1,125 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.network
+
+import java.util.ArrayDeque
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.{Condition, ReentrantLock}
+
+/**
+ * A blocking queue that is prioritization-aware.
+ *
+ * Prioritized elements in the queue get polled before regular elements in the queue.
+ */
+private[network] class PrioritizationAwareBlockingQueue[E](private val capacity: Int) {
```

Comment:
I understand the motivation of this custom queue, but I would really like to avoid it if we can. Let us brainstorm offline. You probably saw [this](http://stackoverflow.com/questions/9588233/multiple-blocking-queues-single-consumer) - seems hacky, but also works.

Reply:
@onurkaraman, what was the conclusion of the offline discussion?

Reply:
The discussion happened 9 months ago, so it's hard to say for sure. From what I remember, we tried just using/composing existing concurrent data structures to get the desired effect but they were all unpleasant hacks.